src/native/eventpipe/README.md

```diff
@@ -0,0 +1,112 @@
+over the RPC channel but there are also others entirely unrelated to EventPipe functionality.
+
+## EventPipe concepts
+
+EventPipe collects structured log messages from instrumented code, potentially buffers it, and then
+emits it to a logging sink, typically a serialized file on disk, named pipe or socket. The schema for
+each logged message is defined in an EventPipeEvent. EventPipeProviders act as namespaces for
+EventPipeEvents. In order to receive log messages a telemetry consumer creates an EventPipeSession
+which records the configuration about which EventPipeEvents should be enabled, which logging
+sink should be used, and manages storage for intermediate storage.
+
+A simple workflow looks like this:
+1. Managed code creates an EventSource. The implementation of EventSource automatically creates
+EventPipeProvider and EventPipeEvent objects that correspond to the EventSource object and the
+individual logging methods on the EventSource. At this point nobody is listening so invoking
+EventPipe WriteEvent() APIs don't record any data.
+
+2. At some point a user uses dotnet-trace to send an IPC command to the runtime that creates
+a new EventPipeSession. This session enables some providers specifying a level and keywords
+which determine the events to enable. The session creates an EventPipeBufferManager to manage
+the in-memory log buffering and starts a thread that will dequeue messages from the buffer
+and serialize them to an outbound stream.
+
+3. Now when EventSource APIs are called, WriteEvent() serializes a buffer of data and saves it
+in the session's buffer. Asynchronously a session specific thread dequeues it, formats it, and
+writes it out to the IPC stream.
+
+4. dotnet-trace listening at the other end of the IPC stream receives the log messages and
+serializes them to disk. Later the user takes that file to Visual Studio or PerfView to visualize
+the contents.
+
+
+## Guide to the code
+
+EventPipe was initially written in C++ for coreclr and then translated to C for use by both Mono and
+CoreCLR runtimes. The code depends solely on the C runtime and a limited set of runtime specific
+implementations for basic datatypes, locks, threads, etc. The set of functionality each runtime is
+expected to provide is defined in ep-rt-* files (ep=EventPipe and rt=Runtime). Each runtime then
+needs to compile a separate lib that implements this ABI and link it together. For example CoreCLR's
+implementation is in ../../coreclr/vm/eventing/eventpipe and mono's implementation is in
+../../mono/mono/eventpipe. Files starting with ep-* are the runtime neutral portions.
+
+Files starting with ds-* are the runtime neutral implementation of DiagnosticServer. ds-rt-* files
+are the runtime specific dependencies, following the same pattern used by EventPipe.
+
+### Getters and Setters
+
+The code uses macros to define getter and setter functions in ep-getter-setter.h. It maps like this:
```

Comment:
It's worth noting that defining these getters and setters is the equivalent of marking the "member" public.  Nothing prevents you from accessing things without a getter/setter though since this is C.